File: src/material.ts
import { HitRecord } from "./hittable";
import { ray, Ray } from "./ray";
import { Ref } from "./ref";
import { randomNum } from "./utils";
import { color, Vec3 } from "./vec3";

export abstract class Material {
  abstract scatter(
    rayIn: Ref<Ray>,
    rec: HitRecord,
    attenuation: Ref<Vec3>,
    scattered: Ref<Ray>,
  ): boolean;
}

export class Lambertian extends Material {
  constructor(public albedo: Vec3) {
    super();
  }

  override scatter(
    _rayIn: Ref<Ray>,
    rec: HitRecord,
    attenuation: Ref<Vec3>,
    scattered: Ref<Ray>,
  ): boolean {
    let scatterDirection = rec.normal!.add(Vec3.randomUnitVector());

    if (scatterDirection.nearZero()) {
      scatterDirection = rec.normal!.clone();
    }

    scattered.value = ray(rec.p!, scatterDirection);
    attenuation.value = this.albedo;
    return true;
  }
}

export class Metal extends Material {
  fuzz: number;
  constructor(
    public albedo: Vec3,
    fuzz: number,
  ) {
    super();

    this.fuzz = fuzz < 1 ? fuzz : 1;
  }

  override scatter(
    rayIn: Ref<Ray>,
    rec: HitRecord,
    attenuation: Ref<Vec3>,
    scattered: Ref<Ray>,
  ): boolean {
    let reflected = Vec3.reflect(rayIn.value!.direction, rec.normal!);
    reflected = reflected.unit.add(Vec3.randomUnitVector().k(this.fuzz));
    scattered.value = ray(rec.p!, reflected);
    attenuation.value = this.albedo;
    return Vec3.dot(scattered.value.direction!, rec.normal!) > 0;
  }
}

export class Dialectric extends Material {
  constructor(public refractionIndex: number) {
    super();
  }

  scatter(
    rayIn: Ref<Ray>,
    rec: HitRecord,
    attenuation: Ref<Vec3>,
    scattered: Ref<Ray>,
  ): boolean {
    attenuation.value = color(1, 1, 1);

    const ri = rec.frontFace!
      ? 1.0 / this.refractionIndex
      : this.refractionIndex;

    const unitDirection = rayIn.value!.direction.unit;

    const cosTheta = Math.min(Vec3.dot(unitDirection.k(-1), rec.normal!), 1.0);
    const sinTheta = Math.sqrt(1 - cosTheta * cosTheta);

    const cannotRefract = ri * sinTheta > 1;

    let direction: Vec3;

    if (cannotRefract || this.reflectance(cosTheta, ri) > randomNum()) {
      direction = Vec3.reflect(unitDirection, rec.normal!);
    } else {
      direction = Vec3.refract(unitDirection, rec.normal!, ri);
    }

    scattered.value = ray(rec.p!, direction);

    return true;
  }

  reflectance(cosine: number, refractionIndex: number) {
    // Schlick's approximation
    let r0 = (1 - refractionIndex) / (1 + refractionIndex);
    r0 = r0 * r0;

    return r0 + (1 - r0) * Math.pow(1 - cosine, 5);
  }
}

File: raytracer/material.go
package raytracer

import (
	"math"
	"math/rand/v2"
	"raytracer/math3"
)

type Material interface {
	Scatter(ray math3.Ray, rec HitRecord) (math3.Vec3, math3.Ray, bool)
}

type Lambertian struct {
	Albedo math3.Vec3
}

func (l Lambertian) Scatter(ray math3.Ray, rec HitRecord) (math3.Vec3, math3.Ray, bool) {
	scatterDir := rec.Normal.Add(math3.RandomUnitVector())
	if scatterDir.IsNearZero() {
		scatterDir = rec.Normal
	}
	return l.Albedo, math3.Ray{Origin: rec.P, Direction: scatterDir}, true
}

type Metal struct {
	Fuzz   float64
	Albedo math3.Vec3
}

func (m Metal) Scatter(ray math3.Ray, rec HitRecord) (math3.Vec3, math3.Ray, bool) {
	reflected := math3.Reflect(ray.Direction, rec.Normal)
	reflected = reflected.Normalize().Add(math3.RandomUnitVector().Scale(m.Fuzz))
	scattered := math3.Ray{Origin: rec.P, Direction: reflected}
	canScatter := math3.Dot(scattered.Direction, rec.Normal) > 0
	return m.Albedo, scattered, canScatter
}

type Dialectric struct {
	RefractionIndex float64
}

func (d Dialectric) Scatter(ray math3.Ray, rec HitRecord) (math3.Vec3, math3.Ray, bool) {
	ri := d.RefractionIndex
	if rec.FrontFace {
		ri = 1 / d.RefractionIndex
	}
	unitDir := ray.Direction.Normalize()
	cosT := math.Min(math3.Dot(unitDir.Scale(-1), rec.Normal), 1)
	sinT := math.Sqrt(math.Max(0.0, 1.0-cosT*cosT))
	cannotRefract := ri*sinT > 1
	var direction math3.Vec3
	if cannotRefract || d.reflectance(cosT, ri) > rand.Float64() {
		direction = math3.Reflect(unitDir, rec.Normal)
	} else {
		direction = math3.Refract(unitDir, rec.Normal, ri)
	}
	scattered := math3.Ray{Origin: rec.P, Direction: direction}
	return math3.Vec3{1, 1, 1}, scattered, true
}

func (d Dialectric) reflectance(cosine float64, refractionIndex float64) float64 {
	r0 := (1 - refractionIndex) / (1 + refractionIndex)
	r0 = r0 * r0
	return r0 + (1-r0)*math.Pow(1-cosine, 5)
}
